File: compiler/tests/org/jetbrains/jet/test/generator/SimpleTestMethodModel.java
/*
 * Copyright 2010-2013 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.jet.test.generator;

import com.intellij.openapi.util.io.FileUtil;
import com.intellij.openapi.util.text.StringUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.jet.JetTestUtils;
import org.jetbrains.jet.utils.Printer;

import java.io.File;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SimpleTestMethodModel implements TestMethodModel {
    private final File rootDir;
    private final File file;
    private final String doTestMethodName;
    private final Pattern filenamePattern;

    public SimpleTestMethodModel(File rootDir, File file, String doTestMethodName, Pattern filenamePattern) {
        this.rootDir = rootDir;
        this.file = file;
        this.doTestMethodName = doTestMethodName;
        this.filenamePattern = filenamePattern;
    }

    @Override
    public void generateBody(@NotNull Printer p, @NotNull String generatorClassFqName) {
        p.println(doTestMethodName, "(\"", JetTestUtils.getFilePath(file), "\");");
    }

    @Override
    public String getDataString() {
        return JetTestUtils.getFilePath(new File(FileUtil.getRelativePath(rootDir, file)));
    }

    @Override
    public String getName() {
        Matcher matcher = filenamePattern.matcher(file.getName());
        boolean found = matcher.find();
        assert found : file.getName() + " isn't matched by regex " + filenamePattern.pattern();
        assert matcher.groupCount() == 1 : filenamePattern.pattern();
        String extractedName = matcher.group(1);

        String unescapedName;
        if (rootDir.equals(file.getParentFile())) {
            unescapedName = extractedName;
        }
        else {
            String relativePath = FileUtil.getRelativePath(rootDir, file.getParentFile());
            unescapedName = relativePath + "-" + StringUtil.capitalize(extractedName);
        }
        return "test" + StringUtil.capitalize(TestGeneratorUtil.escapeForJavaIdentifier(unescapedName));
    }
}

File: generators/test-generator/testFixtures/org/jetbrains/kotlin/generators/model/methods/SimpleTestMethodModel.kt
/*
 * Copyright 2010-2026 JetBrains s.r.o. and Kotlin Programming Language contributors.
 * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
 */

package org.jetbrains.kotlin.generators.model.methods

import com.intellij.openapi.util.io.FileUtil
import org.jetbrains.kotlin.generators.MethodGenerator
import org.jetbrains.kotlin.generators.model.MethodModel
import org.jetbrains.kotlin.generators.model.TestInfraRevision
import org.jetbrains.kotlin.generators.util.TestGeneratorUtil
import org.jetbrains.kotlin.generators.util.getFilePath
import org.jetbrains.kotlin.utils.Printer
import java.io.File
import java.util.regex.Pattern

/**
 * Default model for the test method.
 */
class SimpleTestMethodModel(
    private val testInfraRevision: TestInfraRevision,
    private val rootDir: File,
    val file: File,
    private val filenamePattern: Pattern,
    override val tags: List<String>,
) : MethodModel<SimpleTestMethodModel>() {
    override val generator: MethodGenerator<SimpleTestMethodModel> get() = Generator

    override val dataString: String
        get() {
            val path = FileUtil.getRelativePath(rootDir, file)!!
            return File(path).getFilePath()
        }

    override val name: String
        get() {
            val matcher = filenamePattern.matcher(file.name)
            val found = matcher.find()
            assert(found) { file.name + " isn't matched by regex " + filenamePattern.pattern() }
            assert(matcher.groupCount() >= 1) { filenamePattern.pattern() }
            val extractedName = try {
                matcher.group(1) ?: error("extractedName should not be null: " + filenamePattern.pattern())
            } catch (e: Throwable) {
                throw IllegalStateException("Error generating test ${file.name}", e)
            }
            val unescapedName = if (rootDir == file.parentFile) {
                extractedName
            } else {
                val relativePath = FileUtil.getRelativePath(rootDir, file.parentFile)
                relativePath + "-" + extractedName.replaceFirstChar(Char::uppercaseChar)
            }
            val nameSuffix = TestGeneratorUtil.escapeForJavaIdentifier(unescapedName).replaceFirstChar(Char::uppercaseChar)
            return "test$nameSuffix"
        }

    private object Generator : MethodGenerator<SimpleTestMethodModel>() {
        override fun generateSignature(method: SimpleTestMethodModel, p: Printer) {
            generateDefaultSignature(method, p)
        }

        override fun generateBody(method: SimpleTestMethodModel, p: Printer) {
            val file = method.file
            when (method.testInfraRevision) {
                TestInfraRevision.StandardJUnit5 if file.isFile -> {
                    p.println(RunTestWithDirectoryPrefixMethodModel.METHOD_NAME, "(\"", file.name, "\");")
                }
                else -> {
                    val filePath = file.getFilePath() + if (file.isDirectory) "/" else ""
                    p.println(DEFAULT_RUN_TEST_METHOD_NAME, "(\"", filePath, "\");")
                }
            }
        }
    }
}
